system/Validation/ValidationInterface.php

```diff
@@ -53,7 +53,7 @@ public function check($value, string $rule, array $errors = []): bool;
      *
      * @return ValidationInterface
      */
-    public function withRequest(RequestInterface $request): ValidationInterface;
+    public function withRequest(RequestInterface $request): self;
```

Comment:
I'm not sure about this change for interfaces. I assume PHP handles this correctly (i.e. the implementing class must return a `ValidationInterface` instead of `self`) but it makes it confusing for developers. For example I like to copy-paste an interface when starting a new implementing class and fill in the methods but this will be incorrect if I failed to update the signatures.

Reply:
I'm not sure what you mean, but something like this?
```php
interface ValidationInterface
{
    public function validate(array $data): self;
}

class Validator implements ValidationInterface
{
    private $data;
    public function validate(array $data): ValidationInterface
    {
        $this->data = $data;
        return $this;
    }
}

class Validation implements ValidationInterface
{
    private $data;
    public function validate(array $data): self
    {
        $this->data = $data;
        return $this;
    }
}
```

The second class will fail for 7.3 and below because full return covariance is supported in 7.4+
See https://3v4l.org/bpubF

Reply:
Right. My question is: since the implementing class (Validator, in your example) *must* return `ValidationInterface` instead of `self` would it be cleaner & clearer to keep the class name on the interface itself?

Reply:
It could be done but we'll have to disable the rule entirely as there's no option to skip a rule for a file.

OR

I can defer enabling this rule to when we're requiring 7.4+ so as to support full return covariance.

Reply:
This is always a mess with 7.3 and under. I would be in favor of applying this rule's fixes then disabling the rule and changing these special cases back manually. I'm open to other opinions though.

Reply:
For now, I did what you want. Enabling the effects on classes but not on interfaces, then disabling the fixer. We can revisit this later when we bump PHP requirements.

Reply:
Thanks! I imagine we'll have a ton of changes around typehints when we drop 7.3 so I think that's a great plan for when to revisit.